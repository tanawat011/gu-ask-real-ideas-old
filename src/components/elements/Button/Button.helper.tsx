import clsx from 'clsx'

import { BUTTON_STYLE } from '@constants'
import { PropButton } from '.'

export const renderPrimaryClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-primary-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-primary-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-primary-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-primary-default text-primary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-primary-default text-primary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-primary-default-flat text-primary-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-primary-default', classHover())
  }

  return clsx(
    'bg-primary-default text-white',
    classHover(),
    !isLoading && 'active:bg-primary-pressed',
    hasShadow && 'shadow-nextui shadow-primary-default',
  )
}

export const renderSecondaryClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-secondary-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-secondary-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-secondary-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-secondary-default text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-secondary-default text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx(
      'bg-secondary-default-flat text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-secondary-default', classHover())
  }

  return clsx(
    'bg-secondary-default text-white',
    classHover(),
    !isLoading && 'active:bg-secondary-pressed',
    hasShadow && 'shadow-nextui shadow-secondary-default',
  )
}

export const renderSuccessClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-success-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-success-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-success-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-success-default text-success-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-success-default text-success-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-success-default-flat text-success-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-success-default', classHover())
  }

  return clsx(
    'bg-success-default text-white',
    classHover(),
    !isLoading && 'active:bg-success-pressed',
    hasShadow && 'shadow-nextui shadow-success-default',
  )
}

export const renderWarningClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-warning-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-warning-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-warning-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-warning-default text-warning-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-warning-default text-warning-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-warning-default-flat text-warning-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-warning-default', classHover())
  }

  return clsx(
    'bg-warning-default text-white',
    classHover(),
    !isLoading && 'active:bg-warning-pressed',
    hasShadow && 'shadow-nextui shadow-warning-default',
  )
}

export const renderInfoClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-info-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-info-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-info-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-info-default text-info-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-info-default text-info-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-info-default-flat text-info-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-info-default', classHover())
  }

  return clsx(
    'bg-info-default text-white',
    classHover(),
    !isLoading && 'active:bg-info-pressed',
    hasShadow && 'shadow-nextui shadow-info-default',
  )
}

export const renderDangerClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-danger-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-danger-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-danger-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-danger-default text-danger-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-danger-default text-danger-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-danger-default-flat text-danger-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-danger-default', classHover())
  }

  return clsx(
    'bg-danger-default text-white',
    classHover(),
    !isLoading && 'active:bg-danger-pressed',
    hasShadow && 'shadow-nextui shadow-danger-default',
  )
}
