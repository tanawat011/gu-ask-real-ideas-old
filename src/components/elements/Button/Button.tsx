import { MouseEvent, cloneElement } from 'react'

import {
  BUTTON_STYLE,
  VARIANT,
  LOADING_COLOR,
  LOADING_STYLE,
  SIZE,
} from '@constants'
import { Loading } from '@element/Loader'
import { RegEx } from '@helpers'
import { clsx } from '@libs'

export type PropButton = {
  width?: number
  label?: string
  size?: SIZE
  variant?: VARIANT
  gradients?: [string, string, string] | [string, string] | [string]
  btnStyle?: BUTTON_STYLE
  icon?: React.ReactElement
  iconRight?: React.ReactElement
  iconFit?: boolean
  disabled?: boolean
  loading?: boolean
  rounded?: boolean
  rippleEffect?: boolean
  wAuto?: boolean
  shadow?: boolean
  onClick?: (event: MouseEvent) => void
}

export const Button: React.FC<PropButton> = (props) => {
  const {
    width,
    label,
    size = SIZE.MD,
    variant = VARIANT.PRIMARY,
    gradients,
    icon,
    iconRight,
    btnStyle = BUTTON_STYLE.DEFAULT,
    iconFit = false,
    disabled = false,
    loading = false,
    rounded = false,
    rippleEffect = true,
    wAuto = false,
    shadow,
    onClick,
  } = props

  const createRippleEffect = (event: MouseEvent) => {
    const button = event.currentTarget as EventTarget &
      Element & { offsetLeft: number; offsetTop: number }

    const circle = document.createElement('span')
    const diameter = Math.max(button.clientWidth, button.clientHeight)
    const radius = diameter / 2

    circle.style.width = circle.style.height = `${diameter}px`
    circle.style.left = `${event.clientX - button.offsetLeft - radius}px`
    circle.style.top = `${event.clientY - button.offsetTop - radius}px`
    circle.classList.add('ripple')

    if (btnStyle !== BUTTON_STYLE.DEFAULT) {
      circle.classList.add(`btn__bg-${variant.toLowerCase()}`)
    } else {
      circle.classList.add('bg-black')
    }

    const ripple = button.getElementsByClassName('ripple')[0]

    if (ripple) {
      ripple.remove()
    }

    button.appendChild(circle)
  }

  const isDisabled = () => disabled || loading

  const renderCursor = () => (disabled || loading) && 'cursor-not-allowed'
  const renderDisabled = () => (disabled || loading) && 'btn__disabled'
  const renderRounded = () => rounded && 'btn__rounded'
  const renderAutoWidth = () => wAuto && 'btn__auto-w'
  const renderSize = () => `btn__${size.toLowerCase()}`

  const renderVariant = () => {
    const _variant = variant.toLowerCase()
    const _style = btnStyle.toLowerCase()

    if (disabled) {
      if (btnStyle === BUTTON_STYLE.DEFAULT) {
        return `btn__${_variant}-disabled`
      }

      return `btn__${_variant}-${_style}-disabled`
    }

    if (gradients) {
      const gFirst = `from-${
        RegEx.colorHex.test(gradients[0]) ? `[${gradients[0]}]` : gradients[0]
      }`
      let gSecond = ''
      let gThird = ''

      if (gradients.length === 2) {
        gSecond = `to-${
          RegEx.colorHex.test(gradients[1]) ? `[${gradients[1]}]` : gradients[1]
        }`
      }

      if (gradients.length === 3) {
        gSecond = `via-${
          RegEx.colorHex.test(gradients[1]) ? `[${gradients[1]}]` : gradients[1]
        }`
        gThird = `to-${
          RegEx.colorHex.test(gradients[2]) ? `[${gradients[2]}]` : gradients[2]
        }`
      }

      return clsx('bg-gradient-to-r', gFirst, gSecond, gThird)
    }

    if (btnStyle === BUTTON_STYLE.DEFAULT) {
      return clsx(
        `btn__${_variant}`,
        loading && `btn__${_variant}-loading`,
        shadow && `btn__${_variant}-shadow`,
      )
    }

    return clsx(
      `btn__${_variant}-${_style}`,
      loading && `btn__${_variant}-${_style}-loading`,
    )
  }

  const renderIcon = (
    _icon: React.ReactElement,
    position: 'left' | 'right',
  ) => {
    const isLeft = position === 'left'

    let classPaddingIcon = ''
    let classWidthIcon = label ? 'w-8' : 'w-6'

    if (size === SIZE.XS) {
      classWidthIcon = label ? 'w-6' : 'w-4'
    }

    if (label) {
      classPaddingIcon = isLeft ? 'pr-2' : 'pl-2'
    }

    const classIcon = clsx(
      classWidthIcon,
      classPaddingIcon,
      iconFit && 'absolute',
      iconFit && isLeft ? 'left-4' : 'right-4',
    )

    return cloneElement(_icon, { className: classIcon })
  }

  const renderChild = () => {
    return loading ? (
      <Loading loaderStyle={LOADING_STYLE.DASH} color={LOADING_COLOR.WHITE} />
    ) : (
      <>
        {icon && renderIcon(icon, 'left')}
        {label}
        {iconRight && renderIcon(iconRight, 'right')}
      </>
    )
  }

  return (
    <button
      className={clsx(
        'btn',
        renderCursor(),
        renderDisabled(),
        renderRounded(),
        renderAutoWidth(),
        renderSize(),
        renderVariant(),
      )}
      style={{ ...(width && { width }) }}
      disabled={isDisabled()}
      onClick={(event) => {
        if (rippleEffect) {
          createRippleEffect(event)
        }

        if (onClick) {
          onClick(event)
        }
      }}
    >
      {renderChild()}
    </button>
  )
}
