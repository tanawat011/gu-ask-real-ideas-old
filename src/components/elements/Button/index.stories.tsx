import { ComponentMeta, ComponentStory } from '@storybook/react'

import { Button } from '.'

export default {
  title: 'Button/Default',
  component: Button,
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Default = Template.bind({})

Default.args = {
  title: 'Button',
  size: 'medium',
  color: 'primary-default',
}
