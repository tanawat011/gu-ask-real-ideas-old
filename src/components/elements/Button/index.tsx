import clsx from 'clsx'
import { VFC } from 'react'
import Image from 'next/image'

interface ButtonType {
  title: string
  size: 'small' | 'medium' | 'large'
  color: 'yellow' | 'red' | 'white' | 'bg'
  icon?: string
  isDisabled?: boolean
  width?: number
}

export const Button: VFC<ButtonType> = ({
  width,
  title,
  size,
  color,
  icon,
  isDisabled = false,
}) => {
  const renderSize = (size: string) => {
    if (size === 'small') {
      return 'h-6-5 w-52-7 text-xss rounded'
    }

    if (size === 'small' && color === 'white') {
      return 'h-6 w-28-1 text-xss rounded'
    }

    if (size === 'medium') {
      return 'h-7-5 w-64-2 text-xss rounded-md'
    }

    if (size === 'medium' && color === 'white') {
      return 'h-6 w-64-8 text-xss rounded-md'
    }

    if (size === 'large') {
      return 'h-10 w-64-2 text-xs rounded-lg'
    }

    if (size === 'large' && color === 'white') {
      return 'h-6 w-64-8 text-xs rounded-lg'
    }
  }

  const renderColor = (color: string) => {
    if (color === 'yellow') {
      return clsx(
        'bg-action-default hover:bg-action-hover active:bg-action-pressed',
        isDisabled && 'disabled:bg-action-default opacity-40',
      )
    }

    if (color === 'red') {
      return clsx(
        'bg-secondary-candy hover:bg-candy-hover active:bg-candy-pressed',
        isDisabled && 'disabled:bg-secondary-candy opacity-40',
      )
    }

    if (color === 'white') {
      return clsx(
        'bg-neutral-fade-100 hover:bg-neutral-gray-200 active:bg-neutral-dark-400',
        isDisabled && 'disabled:bg-neutral-fade-100 opacity-40',
      )
    }

    if (color === 'bg') {
      return clsx(
        'bg-neutral-gray-300 hover:bg-neutral-gray-200 active:bg-neutral-dark-400',
        isDisabled && 'disabled:bg-neutral-gray-300 opacity-40',
      )
    }
  }

  return (
    <button
      className={`
        ${renderSize(size)}
        ${renderColor(color)}
        flex items-center justify-center font-medium text-neutral-black
      `}
      style={{ width }}
      disabled={isDisabled}
    >
      {icon && (
        <Image className='w-3 h-3 m-1 pointer-events-none' src={icon} alt='' />
      )}{' '}
      {title}
    </button>
  )
}
