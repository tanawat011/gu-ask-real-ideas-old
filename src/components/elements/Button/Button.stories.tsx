import { ComponentMeta, ComponentStory } from '@storybook/react'

import { Button } from '@element/Button/Button'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    label: {
      name: 'Label',
      table: {
        category: 'Text',
      },
    },
    size: {
      name: 'Size',
      table: {
        category: 'Size',
      },
    },
    color: {
      name: 'Color',
      table: {
        category: 'Colors',
      },
    },
    isLoading: {
      name: 'Set loading',
      table: {
        category: 'Setting',
      },
    },
    isDisabled: {
      name: 'Set disabled',
      table: {
        category: 'Setting',
      },
    },
  },
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Primary = Template.bind({})

Primary.args = {
  label: 'Button',
  size: 'medium',
  color: 'primary',
}
Primary.parameters = {
  docs: {
    description: {
      component: 'Button primary',
    },
  },
}

export const Secondary = Template.bind({})

Secondary.args = {
  label: 'Button',
  size: 'medium',
  color: 'secondary',
}
Secondary.parameters = {
  docs: {
    description: {
      component: 'Button secondary',
    },
  },
}
