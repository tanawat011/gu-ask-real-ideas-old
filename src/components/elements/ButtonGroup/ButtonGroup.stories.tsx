import {
  AdjustmentsIcon,
  AnnotationIcon,
  ArchiveIcon,
} from '@heroicons/react/solid'
import { ComponentMeta, ComponentStory } from '@storybook/react'

import { Button } from '@element/Button'
import { ButtonGroup } from '@element/ButtonGroup'
import { BUTTON_STYLE, SIZE, VARIANT } from '@constants'

const IconMapping = {
  '-': '',
  ArchiveIcon: <ArchiveIcon />,
  AnnotationIcon: <AnnotationIcon />,
  AdjustmentsIcon: <AdjustmentsIcon />,
}
const iconOptions = {
  options: Object.keys(IconMapping),
  mapping: { ...IconMapping },
  control: {
    type: 'select',
  },
}

export default {
  title: 'ButtonGroup',
  component: ButtonGroup,
  parameters: {
    docs: {
      description: {
        component: `Button description
          * In the Docs menu will no effect ripple when click button
        `,
      },
      page: null,
    },
  },
  argTypes: {
    size: {
      control: {
        type: 'select',
      },
    },
    btnStyle: {
      control: {
        type: 'select',
      },
    },
    icon: iconOptions,
    iconRight: iconOptions,
  },
} as ComponentMeta<typeof ButtonGroup>

const Template: ComponentStory<typeof ButtonGroup> = (props) => (
  <>
    <ButtonGroup {...props}>
      <Button label='One' />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' />
      <Button label='Two' />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' />
      <Button label='Two' />
      <Button label='Three' />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' />
      <Button label='Two' />
      <Button label='Three' />
      <Button label='Four' />
    </ButtonGroup>
  </>
)

const TemplateSize: ComponentStory<typeof ButtonGroup> = (props) => {
  return (
    <div className={props.vertical ? 'flex flex-row space-x-4' : ''}>
      {Object.keys(SIZE).map((size, index) => (
        <>
          <ButtonGroup
            key={index}
            {...props}
            size={SIZE[size as keyof typeof SIZE]}
          >
            <Button label='One' />
            <Button label='Two' />
            <Button label='Three' />
          </ButtonGroup>
          <br />
        </>
      ))}
    </div>
  )
}

const TemplateVariant: ComponentStory<typeof ButtonGroup> = (props) => (
  <>
    <ButtonGroup {...props}>
      <Button label='One' variant={VARIANT.DANGER} />
      <Button label='Two' variant={VARIANT.DANGER} />
      <Button label='Three' variant={VARIANT.DANGER} />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button
        label='One'
        variant={VARIANT.SUCCESS}
        btnStyle={BUTTON_STYLE.GHOST}
      />
      <Button
        label='Two'
        variant={VARIANT.SUCCESS}
        btnStyle={BUTTON_STYLE.GHOST}
      />
      <Button
        label='Three'
        variant={VARIANT.SUCCESS}
        btnStyle={BUTTON_STYLE.GHOST}
      />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' variant={VARIANT.SECONDARY} rounded />
      <Button label='Two' variant={VARIANT.SECONDARY} rounded />
      <Button label='Three' variant={VARIANT.SECONDARY} rounded />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' shadow />
      <Button label='Two' shadow />
      <Button label='Three' shadow />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' btnStyle={BUTTON_STYLE.FLAT} />
      <Button label='Two' btnStyle={BUTTON_STYLE.FLAT} />
      <Button label='Three' btnStyle={BUTTON_STYLE.FLAT} />
    </ButtonGroup>
    <br />
    <ButtonGroup {...props}>
      <Button label='One' btnStyle={BUTTON_STYLE.DEFAULT} />
      <Button
        label='Two'
        btnStyle={BUTTON_STYLE.GHOST}
        variant={VARIANT.SECONDARY}
        icon={<ArchiveIcon />}
      />
      <Button
        label='Four'
        btnStyle={BUTTON_STYLE.LIGHT}
        variant={VARIANT.SUCCESS}
      />
      <Button
        label='Three'
        btnStyle={BUTTON_STYLE.FLAT}
        variant={VARIANT.WARNING}
        iconRight={<ArchiveIcon />}
      />
    </ButtonGroup>
  </>
)

export const Default = Template.bind({})

export const Size = TemplateSize.bind({})
Size.args = {
  wAuto: true,
}

export const Shadow = Template.bind({})
Shadow.args = {
  shadow: true,
}

export const Rounded = Template.bind({})
Rounded.args = {
  rounded: true,
}

export const Loading = Template.bind({})
Loading.args = {
  loading: true,
}

export const Disabled = Template.bind({})
Disabled.args = {
  disabled: true,
}

export const OutlineStyle = Template.bind({})
OutlineStyle.args = {
  btnStyle: BUTTON_STYLE.OUTLINE,
}

export const GhostStyle = Template.bind({})
GhostStyle.args = {
  btnStyle: BUTTON_STYLE.GHOST,
}

export const FlatStyle = Template.bind({})
FlatStyle.args = {
  btnStyle: BUTTON_STYLE.FLAT,
}

export const LightStyle = Template.bind({})
LightStyle.args = {
  btnStyle: BUTTON_STYLE.LIGHT,
}

export const Vertical = TemplateSize.bind({})
Vertical.args = {
  vertical: true,
  btnStyle: BUTTON_STYLE.FLAT,
  wAuto: true,
}
export const Variant = TemplateVariant.bind({})
Variant.args = {
  wAuto: true,
}
