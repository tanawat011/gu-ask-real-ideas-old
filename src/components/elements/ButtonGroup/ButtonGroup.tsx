import React from 'react'

import { PropButton } from '@element/Button'
import { clsx } from '@libs'

export type PropButtonGroup = {
  vertical?: boolean
  testId?: string
} & PropButton

export const ButtonGroup: React.FC<PropButtonGroup> = (props) => {
  const { vertical, testId } = props

  const children = React.Children.map(props.children, (child) => {
    if (!React.isValidElement(child)) {
      return child
    }

    const _props = { ...props }

    delete _props.children

    return React.cloneElement(child, {
      ..._props,
      ...child.props,
    })
  })

  return (
    <div
      className={clsx('btn__group', vertical && 'btn__group-vertical')}
      data-testid={testId}
    >
      {children}
    </div>
  )
}
