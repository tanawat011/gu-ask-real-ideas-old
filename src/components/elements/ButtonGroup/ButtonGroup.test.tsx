import { render } from '@testing-library/react'

import { SIZE, VARIANT } from '@constants'
import { Button } from '@element/Button'
import { ButtonGroup } from '@element/ButtonGroup'

describe('<ButtonGroup />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render to match snapshot', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <ButtonGroup>
        <Button label='One' />
      </ButtonGroup>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render to match snapshot and to be vertical group button and', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <ButtonGroup vertical>
        <Button label='One' />
      </ButtonGroup>,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should be group button, default button', async () => {
    expect.assertions(2)

    const { findByTestId, findAllByRole } = render(
      <ButtonGroup testId='test'>
        <Button label='One' />
      </ButtonGroup>,
    )

    const div = await findByTestId('test')
    const buttons = await findAllByRole('button')

    expect(div).toHaveClass(`btn__group`)
    expect(buttons[0]).toHaveClass(`btn btn__md btn__primary`)
  })

  it('should be group button, all default button', async () => {
    expect.assertions(2)

    const { findAllByRole } = render(
      <ButtonGroup>
        <Button label='One' />
        <Button label='Two' />
      </ButtonGroup>,
    )

    const buttons = await findAllByRole('button')

    expect(buttons[0]).toHaveClass(`btn btn__md btn__primary`)
    expect(buttons[1]).toHaveClass(`btn btn__md btn__primary`)
  })

  it('should be vertical group button', async () => {
    expect.assertions(1)

    const { findByTestId } = render(
      <ButtonGroup testId='test' vertical>
        <Button label='One' />
        <Button label='Two' />
      </ButtonGroup>,
    )

    const div = await findByTestId('test')

    expect(div).toHaveClass(`btn__group btn__group-vertical`)
  })

  it('should be group button, child is text', async () => {
    expect.assertions(1)

    const { findByTestId } = render(
      <ButtonGroup testId='test'>xxx</ButtonGroup>,
    )

    const div = await findByTestId('test')

    expect(div).toHaveClass(`btn__group`)
  })

  it('should be group button, no child', async () => {
    expect.assertions(1)

    const { findByTestId } = render(<ButtonGroup testId='test' />)

    const div = await findByTestId('test')

    expect(div).toHaveClass(`btn__group`)
  })
})
