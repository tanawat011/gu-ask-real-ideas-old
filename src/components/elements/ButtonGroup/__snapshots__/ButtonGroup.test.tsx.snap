// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> should render to match snapshot 1`] = `
<DocumentFragment>
  <div
    class="btn__group"
  >
    <button
      class="btn btn__md btn__primary"
    >
      One
    </button>
  </div>
</DocumentFragment>
`;

exports[`<ButtonGroup /> should render to match snapshot and to be vertical group button and 1`] = `
<DocumentFragment>
  <div
    class="btn__group btn__group-vertical"
  >
    <button
      class="btn btn__md btn__primary"
    >
      One
    </button>
  </div>
</DocumentFragment>
`;
